Crash in v1.1.0 using standard_d2s_v3 VMs

The app insights uploader crashes with the following stacktrace

Using v1.1.0, VMs are all standard_d2s_v3

File path: D:\batch\tasks\startup\wd\node-stats.err.log

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x0 pc=0x660465]

goroutine 1 [running]:
github.com/Azure/batch-insights/pkg.AppInsightsService.UploadStats(0x788a20, 0xc0000058a0, 0xc0000ee000, 0x0, 0x0, 0x0, 0xc000014ca0, 0x2, 0x2, 0x0, ...)
	D:/a/1/s/pkg/appinsights.go:43 +0x495
github.com/Azure/batch-insights/pkg.ListenForStats(0xc000024090, 0xb, 0xc00000a150, 0x26, 0xc00000a180, 0x24)
	D:/a/1/s/pkg/batchinsights.go:73 +0x3f4
main.main()
	D:/a/1/s/main.go:32 +0x321
```

--- a/pkg/appinsights.go
+++ b/pkg/appinsights.go
@@ -43,13 +43,13 @@ func (service AppInsightsService) UploadStats(stats NodeStats) {
 		client.TrackMetric("Memory available", float64(stats.memory.Total-stats.memory.Used))
 	}
 	if stats.diskIO != nil {
-		client.TrackMetric("Disk read", float64(stats.diskIO.readBps))
-		client.TrackMetric("Disk write", float64(stats.diskIO.writeBps))
+		client.TrackMetric("Disk read", float64(stats.diskIO.ReadBps))
+		client.TrackMetric("Disk write", float64(stats.diskIO.WriteBps))
 	}
 
 	if stats.netIO != nil {
-		client.TrackMetric("Network read", float64(stats.netIO.readBps))
-		client.TrackMetric("Network write", float64(stats.netIO.writeBps))
+		client.TrackMetric("Network read", float64(stats.netIO.ReadBps))
+		client.TrackMetric("Network write", float64(stats.netIO.WriteBps))
 	}
 
 	if len(stats.gpus) > 0 {

--- a/pkg/appinsights.go
+++ b/pkg/appinsights.go
@@ -38,12 +38,19 @@ func (service AppInsightsService) UploadStats(stats NodeStats) {
 		client.Track(freeMetric)
 	}
 
-	client.TrackMetric("Memory used", float64(stats.memory.Used))
-	client.TrackMetric("Memory available", float64(stats.memory.Total-stats.memory.Used))
-	client.TrackMetric("Disk read", float64(stats.diskIO.readBps))
-	client.TrackMetric("Disk write", float64(stats.diskIO.writeBps))
-	client.TrackMetric("Network read", float64(stats.netIO.readBps))
-	client.TrackMetric("Network write", float64(stats.netIO.writeBps))
+	if stats.memory != nil {
+		client.TrackMetric("Memory used", float64(stats.memory.Used))
+		client.TrackMetric("Memory available", float64(stats.memory.Total-stats.memory.Used))
+	}
+	if stats.diskIO != nil {
+		client.TrackMetric("Disk read", float64(stats.diskIO.readBps))
+		client.TrackMetric("Disk write", float64(stats.diskIO.writeBps))
+	}
+
+	if stats.netIO != nil {
+		client.TrackMetric("Network read", float64(stats.netIO.readBps))
+		client.TrackMetric("Network write", float64(stats.netIO.writeBps))
+	}
 
 	if len(stats.gpus) > 0 {
 		for cpuN, usage := range stats.gpus {

--- a/pkg/batchinsights.go
+++ b/pkg/batchinsights.go
@@ -2,51 +2,21 @@ package batchinsights
 
 import (
 	"fmt"
-	"os"
 	"runtime"
 	"time"
 
-	"github.com/dustin/go-humanize"
-
 	"github.com/Azure/batch-insights/pkg/cpu"
-	"github.com/shirou/gopsutil/disk"
+	"github.com/Azure/batch-insights/pkg/disk"
+	"github.com/Azure/batch-insights/pkg/utils"
+	"github.com/dustin/go-humanize"
 	"github.com/shirou/gopsutil/mem"
 	"github.com/shirou/gopsutil/net"
 )
 
-var IS_PLATFORM_WINDOWS = runtime.GOOS == "windows"
-
 const STATS_POLL_RATE = time.Duration(5) * time.Second
 
-func pathExists(path string) (bool, error) {
-	_, err := os.Stat(path)
-	if err == nil {
-		return true, nil
-	}
-	if os.IsNotExist(err) {
-		return false, nil
-	}
-	return true, err
-}
-
-func getDiskToWatch() []string {
-	if IS_PLATFORM_WINDOWS == true {
-		return []string{"C:/", "D:/"}
-	} else {
-		var osDisk = "/"
-		var userDisk = "/mnt/resources"
-		var exists, _ = pathExists(userDisk)
-
-		if !exists {
-			userDisk = "/mnt"
-		}
-		return []string{osDisk, userDisk}
-	}
-}
-
 func ListenForStats(poolId string, nodeId string, appInsightsKey string) {
-	var diskIO = IOAggregator{}
-	var netIO = IOAggregator{}
+	var netIO = utils.IOAggregator{}
 	var gpuStatsCollector = NewGPUStatsCollector()
 	defer gpuStatsCollector.Shutdown()
 
@@ -63,8 +33,8 @@ func ListenForStats(poolId string, nodeId string, appInsightsKey string) {
 		var stats = NodeStats{
 			memory:      v,
 			cpuPercents: cpus,
-			diskUsage:   getDiskUsage(),
-			diskIO:      getDiskIO(&diskIO),
+			diskUsage:   disk.GetDiskUsage(),
+			diskIO:      disk.DiskIO(),
 			netIO:       getNetIO(&netIO),
 			gpus:        gpuStatsCollector.GetStats(),
 		}
@@ -77,26 +47,7 @@ func ListenForStats(poolId string, nodeId string, appInsightsKey string) {
 	}
 }
 
-func getDiskIO(diskIO *IOAggregator) *IOStats {
-	var counters, err = disk.IOCounters()
-
-	if err != nil {
-		fmt.Println("Error while retrieving Disk IO", err)
-		return nil
-	}
-	var readBytes uint64 = 0
-	var writeBytes uint64 = 0
-
-	for _, v := range counters {
-		readBytes += v.ReadBytes
-		writeBytes += v.WriteBytes
-	}
-	var stats = diskIO.UpdateAggregates(readBytes, writeBytes)
-	fmt.Println("stats", stats);
-	return &stats
-}
-
-func getNetIO(diskIO *IOAggregator) *IOStats {
+func getNetIO(diskIO *utils.IOAggregator) *utils.IOStats {
 	var counters, err = net.IOCounters(false)
 
 	if err != nil {
@@ -106,22 +57,6 @@ func getNetIO(diskIO *IOAggregator) *IOStats {
 		return &stats
 	}
 	return nil
-}
-
-func getDiskUsage() []*disk.UsageStat {
-	var disks = getDiskToWatch()
-	var stats []*disk.UsageStat
-
-	for _, diskPath := range disks {
-		usage, err := disk.Usage(diskPath)
-		if err == nil {
-			stats = append(stats, usage)
-		} else {
-			fmt.Println(err)
-		}
-	}
-
-	return stats
 }
 
 func PrintSystemInfo() {
@@ -143,11 +78,11 @@ func printStats(stats NodeStats) {
 	}
 
 	if stats.diskIO != nil {
-		fmt.Printf("Disk IO: R:%sps, W:%sps\n", humanize.Bytes(stats.diskIO.readBps), humanize.Bytes(stats.diskIO.writeBps))
+		fmt.Printf("Disk IO: R:%sps, W:%sps\n", humanize.Bytes(stats.diskIO.ReadBps), humanize.Bytes(stats.diskIO.WriteBps))
 	}
 
 	if stats.netIO != nil {
-		fmt.Printf("NET IO: R:%sps, S:%sps\n", humanize.Bytes(stats.netIO.readBps), humanize.Bytes(stats.netIO.writeBps))
+		fmt.Printf("NET IO: R:%sps, S:%sps\n", humanize.Bytes(stats.netIO.ReadBps), humanize.Bytes(stats.netIO.WriteBps))
 	}
 
 	if len(stats.gpus) > 0 {

--- a/pkg/batchinsights.go
+++ b/pkg/batchinsights.go
@@ -92,6 +92,7 @@ func getDiskIO(diskIO *IOAggregator) *IOStats {
 		writeBytes += v.WriteBytes
 	}
 	var stats = diskIO.UpdateAggregates(readBytes, writeBytes)
+	fmt.Println("stats", stats);
 	return &stats
 }
 

--- a/pkg/batchinsights.go
+++ b/pkg/batchinsights.go
@@ -81,7 +81,7 @@ func getDiskIO(diskIO *IOAggregator) *IOStats {
 	var counters, err = disk.IOCounters()
 
 	if err != nil {
-		fmt.Println(err)
+		fmt.Println("Error while retrieving Disk IO", err)
 		return nil
 	}
 	var readBytes uint64 = 0

--- a/pkg/cpu/cpu_windows.go
+++ b/pkg/cpu/cpu_windows.go
@@ -4,11 +4,8 @@ package cpu
 
 import (
 	"context"
-	"github.com/StackExchange/wmi"
-	"time"
+	"github.com/Azure/batch-insights/pkg/wmi"
 )
-
-const Timeout = 3 * time.Second
 
 type CPUStat struct {
 	value     uint64
@@ -65,30 +62,10 @@ func perfInfoWithContext(ctx context.Context) ([]win32_PerfRawData_Counters_Proc
 	var ret []win32_PerfRawData_Counters_ProcessorInformation
 
 	q := wmi.CreateQuery(&ret, "WHERE NOT Name LIKE '%_Total'")
-	err := WMIQueryWithContext(ctx, q, &ret)
+	err := wmi.QueryWithContext(ctx, q, &ret)
 	if err != nil {
 		return []win32_PerfRawData_Counters_ProcessorInformation{}, err
 	}
 
 	return ret, err
-}
-
-func WMIQueryWithContext(ctx context.Context, query string, dst interface{}, connectServerArgs ...interface{}) error {
-	if _, ok := ctx.Deadline(); !ok {
-		ctxTimeout, cancel := context.WithTimeout(ctx, Timeout)
-		defer cancel()
-		ctx = ctxTimeout
-	}
-
-	errChan := make(chan error, 1)
-	go func() {
-		errChan <- wmi.Query(query, dst, connectServerArgs...)
-	}()
-
-	select {
-	case <-ctx.Done():
-		return ctx.Err()
-	case err := <-errChan:
-		return err
-	}
 }

--- a/pkg/disk/disk.go
+++ b/pkg/disk/disk.go
@@ -1,0 +1,53 @@
+package disk
+
+import (
+	"fmt"
+	"os"
+	"runtime"
+
+	psutils_disk "github.com/shirou/gopsutil/disk"
+)
+
+var IS_PLATFORM_WINDOWS = runtime.GOOS == "windows"
+
+func GetDiskUsage() []*psutils_disk.UsageStat {
+	var disks = getDiskToWatch()
+	var stats []*psutils_disk.UsageStat
+
+	for _, diskPath := range disks {
+		usage, err := psutils_disk.Usage(diskPath)
+		if err == nil {
+			stats = append(stats, usage)
+		} else {
+			fmt.Println(err)
+		}
+	}
+
+	return stats
+}
+
+func getDiskToWatch() []string {
+	if IS_PLATFORM_WINDOWS == true {
+		return []string{"C:/", "D:/"}
+	} else {
+		var osDisk = "/"
+		var userDisk = "/mnt/resources"
+		var exists, _ = pathExists(userDisk)
+
+		if !exists {
+			userDisk = "/mnt"
+		}
+		return []string{osDisk, userDisk}
+	}
+}
+
+func pathExists(path string) (bool, error) {
+	_, err := os.Stat(path)
+	if err == nil {
+		return true, nil
+	}
+	if os.IsNotExist(err) {
+		return false, nil
+	}
+	return true, err
+}

--- a/pkg/disk/disk_linux.go
+++ b/pkg/disk/disk_linux.go
@@ -1,0 +1,31 @@
+// +build linux
+
+package disk
+
+import (
+	"fmt"
+
+	"github.com/Azure/batch-insights/pkg/utils"
+	psutils_disk "github.com/shirou/gopsutil/disk"
+)
+
+var diskIO = utils.IOAggregator{}
+
+func DiskIO() *utils.IOStats {
+	var counters, err = psutils_disk.IOCounters()
+
+	if err != nil {
+		fmt.Println("Error while retrieving Disk IO", err)
+		return nil
+	}
+	var readBytes uint64 = 0
+	var writeBytes uint64 = 0
+
+	for _, v := range counters {
+		fmt.Println("stats", v.WriteBytes, v.WriteTime, v.WriteCount)
+		readBytes += v.ReadBytes
+		writeBytes += v.WriteBytes
+	}
+	var stats = diskIO.UpdateAggregates(readBytes, writeBytes)
+	return &stats
+}

--- a/pkg/disk/disk_windows.go
+++ b/pkg/disk/disk_windows.go
@@ -1,0 +1,50 @@
+// +build windows
+
+package disk
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/batch-insights/pkg/utils"
+	"github.com/Azure/batch-insights/pkg/wmi"
+)
+
+type win32_PerfRawData_PerfDisk_PhysicalDisk struct {
+	Name                      string
+	AvgDiskBytesPerRead       uint64
+	AvgDiskBytesPerRead_Base  uint64
+	AvgDiskBytesPerWrite      uint64
+	AvgDiskBytesPerWrite_Base uint64
+	AvgDiskReadQueueLength    uint64
+	AvgDiskWriteQueueLength   uint64
+	AvgDisksecPerRead         uint64
+	AvgDisksecPerWrite        uint64
+}
+
+var diskIO = utils.IOAggregator{}
+
+func DiskIO() *utils.IOStats {
+	return DiskIOWithContext(context.Background())
+}
+
+func DiskIOWithContext(ctx context.Context, names ...string) *utils.IOStats {
+	var ret []win32_PerfRawData_PerfDisk_PhysicalDisk
+
+	q := wmi.CreateQuery(&ret, "WHERE NOT Name LIKE '%_Total'")
+	err := wmi.QueryWithContext(ctx, q, &ret)
+	if err != nil {
+		fmt.Println("Error while retrieving DISK IO", err)
+		return nil
+	}
+
+	var readBytes uint64 = 0
+	var writeBytes uint64 = 0
+	for _, v := range ret {
+		readBytes += v.AvgDiskBytesPerRead
+		writeBytes += v.AvgDiskBytesPerWrite
+	}
+	stats := diskIO.UpdateAggregates(readBytes, writeBytes)
+
+	return &stats
+}

--- a/pkg/node_stats.go
+++ b/pkg/node_stats.go
@@ -3,13 +3,15 @@ package batchinsights
 import (
 	"github.com/shirou/gopsutil/disk"
 	"github.com/shirou/gopsutil/mem"
+
+	"github.com/Azure/batch-insights/pkg/utils"
 )
 
 type NodeStats struct {
 	memory      *mem.VirtualMemoryStat
 	cpuPercents []float64
 	diskUsage   []*disk.UsageStat
-	diskIO      *IOStats
-	netIO       *IOStats
+	diskIO      *utils.IOStats
+	netIO       *utils.IOStats
 	gpus        []GPUUsage
 }

--- a/pkg/utils/io_aggregator.go
+++ b/pkg/utils/io_aggregator.go
@@ -1,4 +1,4 @@
-package batchinsights
+package utils
 
 import (
 	"time"
@@ -11,8 +11,8 @@ type IOAggregator struct {
 }
 
 type IOStats struct {
-	readBps  uint64
-	writeBps uint64
+	ReadBps  uint64
+	WriteBps uint64
 }
 
 func (aggregator *IOAggregator) UpdateAggregates(currentRead uint64, currentWrite uint64) IOStats {
@@ -32,7 +32,7 @@ func (aggregator *IOAggregator) UpdateAggregates(currentRead uint64, currentWrit
 	aggregator.lastWrite = currentWrite
 
 	return IOStats{
-		readBps:  readBps,
-		writeBps: writeBps,
+		ReadBps:  readBps,
+		WriteBps: writeBps,
 	}
 }

--- a/pkg/wmi/wmi.go
+++ b/pkg/wmi/wmi.go
@@ -1,0 +1,34 @@
+package wmi
+
+import (
+	"context"
+	"time"
+
+	wmi_lib "github.com/StackExchange/wmi"
+)
+
+const Timeout = 3 * time.Second
+
+func CreateQuery(src interface{}, where string) string {
+	return wmi_lib.CreateQuery(src, where)
+}
+
+func QueryWithContext(ctx context.Context, query string, dst interface{}, connectServerArgs ...interface{}) error {
+	if _, ok := ctx.Deadline(); !ok {
+		ctxTimeout, cancel := context.WithTimeout(ctx, Timeout)
+		defer cancel()
+		ctx = ctxTimeout
+	}
+
+	errChan := make(chan error, 1)
+	go func() {
+		errChan <- wmi_lib.Query(query, dst, connectServerArgs...)
+	}()
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case err := <-errChan:
+		return err
+	}
+}

--- a/scripts/dev-windows.ps1
+++ b/scripts/dev-windows.ps1
@@ -1,12 +1,13 @@
 $ErrorActionPreference = "Stop"
 
 $wd = $env:AZ_BATCH_TASK_WORKING_DIR
+$branch = $env:BATCH_INSIGHTS_BRANCH
 
 Invoke-Expression ((New-Object System.Net.WebClient).DownloadString('https://chocolatey.org/install.ps1'))
 choco install -y golang git
 $env:Path = [System.Environment]::GetEnvironmentVariable("Path","Machine") + ";" + [System.Environment]::GetEnvironmentVariable("Path","User")
 
-git clone https://github.com/Azure/batch-insights
+git clone https://github.com/Azure/batch-insights -b $branch
 
 Set-Location ./batch-insights
 

--- a/scripts/dev-windows.ps1
+++ b/scripts/dev-windows.ps1
@@ -4,7 +4,7 @@ $wd = $env:AZ_BATCH_TASK_WORKING_DIR
 $branch = $env:BATCH_INSIGHTS_BRANCH
 
 Invoke-Expression ((New-Object System.Net.WebClient).DownloadString('https://chocolatey.org/install.ps1'))
-choco install -y golang git
+choco install -y golang git mingw
 $env:Path = [System.Environment]::GetEnvironmentVariable("Path","Machine") + ";" + [System.Environment]::GetEnvironmentVariable("Path","User")
 
 git clone https://github.com/Azure/batch-insights -b $branch

